source/simulator/helper_functions.cc

```diff
@@ -2403,6 +2404,260 @@ namespace aspect
+
+    // Get the already-computed dynamic topography solution.
+    const LinearAlgebra::BlockVector topo_vector = dynamic_topography.topography_vector();
+    std::vector<double> topo_values( quadrature_formula.size() );
+
+    // Get the density_above from the dynamic topography postprocessor
+    const double density_above = dynamic_topography.return_density_above();
+
+    for (; cell!=endc; ++cell)
+      if (cell->is_locally_owned())
+        {
+          fe_values.reinit (cell);
+
+          // Evaluate the material properties and the solution within the cell
+          MaterialModel::MaterialModelInputs<dim> in(fe_values, cell, introspection, solution);
+          MaterialModel::MaterialModelOutputs<dim> out(fe_values.n_quadrature_points, introspection.n_compositional_fields);
+          material_model->evaluate(in, out);
+
+          MaterialModel::MaterialModelInputs<dim> in_adjoint(fe_values, cell, introspection, current_adjoint_solution);
+
+          // Include model averaging. The assembler includes this as well because it gets a precomputed material model inputs object
+          // Note that material properties are averaged here but not in the surface integral further down because that
+          // functionality is not yet available.
+          MaterialModel::MaterialAveraging::average (parameters.material_averaging,
+                                                     cell,
+                                                     fe_values.get_quadrature(),
+                                                     fe_values.get_mapping(),
+                                                     out);
+
+          local_rhs = 0.;
+          local_mass_matrix = 0.;
+
+
+          // Assemble the volume term
+
+          for (unsigned int q=0; q<n_q_points; ++q)
+            {
+              // Get parameters for the density kernel
+              const Tensor<1,dim> velocity_adjoint = in_adjoint.velocity[q];
+              const Tensor<1,dim> gravity = gravity_model->gravity_vector(in.position[q]);
+
+              // Get parameters for the viscosity kernel
+              const double eta = out.viscosities[q];
+              const SymmetricTensor<2,dim> strain_rate_forward = in.strain_rate[q] - 1./3 * trace(in.strain_rate[q]) * unit_symmetric_tensor<dim>();
+              const SymmetricTensor<2,dim> strain_rate_adjoint = in_adjoint.strain_rate[q] - 1./3 * trace(in_adjoint.strain_rate[q]) * unit_symmetric_tensor<dim>();
+
+              for (unsigned int k=0; k<dofs_per_cell; ++k)
+                {
+                  phi_rho[k] = fe_values[introspection.extractors.compositional_fields[density_idx]].value(k,q);
+                  phi_eta[k] = fe_values[introspection.extractors.compositional_fields[viscosity_idx]].value(k,q);
+                  // we will need the values of the 0th compositional field for the mass matrix assembly below.
+                  phi_0[k] = fe_values[introspection.extractors.compositional_fields[0]].value(k,q);
```

Comment:
I believe I need the first compositional field to assemble the mass matrix in line 2537. I'm not exactly sure why but if it's not the first field it doesn't work. I'm also doing something similar in line 2641 with the zero_comp_block variable. Since I'm not sure why it needs to be the first field I'm not sure what a good symbolic name is.

Reply:
What happens when you pick a different compositional field? Is it the case that we simply didn't set up the sparsity pattern for this block, as discussed in a different comment? That would be the place to fix it.

Reply:
If I change this to picking the 1st (instead of the 0th compositional field) I get the following run time error

An error occurred in line <1684> of file </var/folders/8z/hlb6vc015qjggytkxn84m6_c0000gn/T/heltai/spack-stage/spack-stage-dealii-9.3.0-zy7k3uwnakcqjvrajvacy5l4jrl7eaex/spack-src/source/lac/trilinos_sparse_matrix.cc> in function
    void dealii::TrilinosWrappers::SparseMatrix::add(const dealii::TrilinosWrappers::SparseMatrix::size_type, const dealii::TrilinosWrappers::SparseMatrix::size_type, const dealii::TrilinosWrappers::SparseMatrix::size_type *, const dealii::TrilinosScalar *, const bool, const bool)
The violated condition was: 
    ierr == 0
Additional information: 
    An error with error number 2 occurred while calling a Trilinos
    function

Yes maybe there is an error with the sparsity pattern for this block. Maybe we could talk about that?

Reply:
As mentioned in the other comment, let that go for now. Just use the (comp0,comp0) block of the matrix.